File: packages/url_launcher/url_launcher_android/android/src/test/java/io/flutter/plugins/urllauncher/UrlLauncherTest.java
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.urllauncher;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertThrows;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.isNull;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import android.app.Activity;
import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.provider.Browser;
import androidx.browser.customtabs.CustomTabsIntent;
import androidx.test.core.app.ApplicationProvider;
import java.util.HashMap;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

@RunWith(RobolectricTestRunner.class)
public class UrlLauncherTest {
  @Test
  public void canLaunch_createsIntentWithPassedUrl() {
    UrlLauncher.IntentResolver resolver = mock(UrlLauncher.IntentResolver.class);
    UrlLauncher api = new UrlLauncher(ApplicationProvider.getApplicationContext(), resolver);
    Uri url = Uri.parse("https://flutter.dev");
    when(resolver.getHandlerComponentName(any())).thenReturn(null);

    api.canLaunchUrl(url.toString());

    final ArgumentCaptor<Intent> intentCaptor = ArgumentCaptor.forClass(Intent.class);
    verify(resolver).getHandlerComponentName(intentCaptor.capture());
    assertEquals(url, intentCaptor.getValue().getData());
  }

  @Test
  public void canLaunch_returnsTrue() {
    UrlLauncher api =
        new UrlLauncher(ApplicationProvider.getApplicationContext(), intent -> "some.component");

    Boolean result = api.canLaunchUrl("https://flutter.dev");

    assertTrue(result);
  }

  @Test
  public void canLaunch_returnsFalse() {
    UrlLauncher api = new UrlLauncher(ApplicationProvider.getApplicationContext(), intent -> null);

    Boolean result = api.canLaunchUrl("https://flutter.dev");

    assertFalse(result);
  }

  // Integration testing on emulators won't work as expected without the workaround this tests
  // for, since it will be returned even for intentionally bogus schemes.
  @Test
  public void canLaunch_returnsFalseForEmulatorFallbackComponent() {
    UrlLauncher api =
        new UrlLauncher(
            ApplicationProvider.getApplicationContext(),
            intent -> "{com.android.fallback/com.android.fallback.Fallback}");

    Boolean result = api.canLaunchUrl("https://flutter.dev");

    assertFalse(result);
  }

  @Test
  public void launch_throwsForNoCurrentActivity() {
    UrlLauncher api = new UrlLauncher(ApplicationProvider.getApplicationContext());
    api.setActivity(null);

    Messages.FlutterError exception =
        assertThrows(
            Messages.FlutterError.class,
            () -> api.launchUrl("https://flutter.dev", new HashMap<>(), false));
    assertEquals("NO_ACTIVITY", exception.code);
  }

  @Test
  public void launch_createsIntentWithPassedUrl() {
    Activity activity = mock(Activity.class);
    String url = "https://flutter.dev";
    UrlLauncher api = new UrlLauncher(ApplicationProvider.getApplicationContext());
    api.setActivity(activity);
    doThrow(new ActivityNotFoundException()).when(activity).startActivity(any());

    api.launchUrl("https://flutter.dev", new HashMap<>(), false);

    final ArgumentCaptor<Intent> intentCaptor = ArgumentCaptor.forClass(Intent.class);
    verify(activity).startActivity(intentCaptor.capture());
    assertEquals(url, intentCaptor.getValue().getData().toString());
    assertEquals(0, intentCaptor.getValue().getFlags() & Intent.FLAG_ACTIVITY_REQUIRE_NON_BROWSER);
  }

  @Config(minSdk = 30)
  @Test
  public void launch_setsRequireNonBrowserWhenRequested() {
    Activity activity = mock(Activity.class);
    String url = "https://flutter.dev";
    UrlLauncher api = new UrlLauncher(ApplicationProvider.getApplicationContext());
    api.setActivity(activity);
    doThrow(new ActivityNotFoundException()).when(activity).startActivity(any());

    api.launchUrl("https://flutter.dev", new HashMap<>(), true);

    final ArgumentCaptor<Intent> intentCaptor = ArgumentCaptor.forClass(Intent.class);
    verify(activity).startActivity(intentCaptor.capture());
    assertEquals(
        Intent.FLAG_ACTIVITY_REQUIRE_NON_BROWSER,
        intentCaptor.getValue().getFlags() & Intent.FLAG_ACTIVITY_REQUIRE_NON_BROWSER);
  }

  @Test
  public void launch_returnsFalse() {
    Activity activity = mock(Activity.class);
    UrlLauncher api = new UrlLauncher(ApplicationProvider.getApplicationContext());
    api.setActivity(activity);
    doThrow(new ActivityNotFoundException()).when(activity).startActivity(any());

    boolean result = api.launchUrl("https://flutter.dev", new HashMap<>(), false);

    assertFalse(result);
  }

  @Test
  public void launch_returnsTrue() {
    Activity activity = mock(Activity.class);
    UrlLauncher api = new UrlLauncher(ApplicationProvider.getApplicationContext());
    api.setActivity(activity);

    boolean result = api.launchUrl("https://flutter.dev", new HashMap<>(), false);

    assertTrue(result);
  }

  @Test
  public void openUrlInApp_opensUrlInWebViewIfNecessary() {
    Activity activity = mock(Activity.class);
    UrlLauncher api = new UrlLauncher(ApplicationProvider.getApplicationContext());
    api.setActivity(activity);
    String url = "https://flutter.dev";
    boolean enableJavaScript = false;
    boolean enableDomStorage = false;
    HashMap<String, String> headers = new HashMap<>();
    headers.put("key", "value");
    boolean showTitle = false;

    boolean result =
        api.openUrlInApp(
            url,
            true,
            new Messages.WebViewOptions.Builder()
                .setEnableJavaScript(enableJavaScript)
                .setEnableDomStorage(enableDomStorage)
                .setHeaders(headers)
                .build(),
            new Messages.BrowserOptions.Builder().setShowTitle(showTitle).build());

    final ArgumentCaptor<Intent> intentCaptor = ArgumentCaptor.forClass(Intent.class);
    verify(activity).startActivity(intentCaptor.capture());
    assertTrue(result);
    assertEquals(url, intentCaptor.getValue().getExtras().getString(WebViewActivity.URL_EXTRA));
    assertEquals(
        enableJavaScript,
        intentCaptor.getValue().getExtras().getBoolean(WebViewActivity.ENABLE_JS_EXTRA));
    assertEquals(
        enableDomStorage,
        intentCaptor.getValue().getExtras().getBoolean(WebViewActivity.ENABLE_DOM_EXTRA));
  }

  @Test
  public void openWebView_opensUrlInWebViewIfRequested() {
    Activity activity = mock(Activity.class);
    UrlLauncher api = new UrlLauncher(ApplicationProvider.getApplicationContext());
    api.setActivity(activity);
    String url = "https://flutter.dev";

    boolean result =
        api.openUrlInApp(
            url,
            false,
            new Messages.WebViewOptions.Builder()
                .setEnableJavaScript(false)
                .setEnableDomStorage(false)
                .setHeaders(new HashMap<>())
                .build(),
            new Messages.BrowserOptions.Builder().setShowTitle(true).build());

    final ArgumentCaptor<Intent> intentCaptor = ArgumentCaptor.forClass(Intent.class);
    verify(activity).startActivity(intentCaptor.capture());
    assertTrue(result);
    assertEquals(url, intentCaptor.getValue().getExtras().getString(WebViewActivity.URL_EXTRA));
  }

  @Test
  public void openWebView_opensUrlInCustomTabs() {
    Activity activity = mock(Activity.class);
    UrlLauncher api = new UrlLauncher(ApplicationProvider.getApplicationContext());
    api.setActivity(activity);
    String url = "https://flutter.dev";

    boolean result =
        api.openUrlInApp(
            url,
            true,
            new Messages.WebViewOptions.Builder()
                .setEnableJavaScript(false)
                .setEnableDomStorage(false)
                .setHeaders(new HashMap<>())
                .build(),
            new Messages.BrowserOptions.Builder().setShowTitle(false).build());

    final ArgumentCaptor<Intent> intentCaptor = ArgumentCaptor.forClass(Intent.class);
    verify(activity).startActivity(intentCaptor.capture(), isNull());
    assertTrue(result);
    assertEquals(Intent.ACTION_VIEW, intentCaptor.getValue().getAction());
    assertNull(intentCaptor.getValue().getComponent());
  }

  @Test
  public void openWebView_opensUrlInCustomTabsWithCORSAllowedHeader() {
    Activity activity = mock(Activity.class);
    UrlLauncher api = new UrlLauncher(ApplicationProvider.getApplicationContext());
    api.setActivity(activity);
    String url = "https://flutter.dev";
    HashMap<String, String> headers = new HashMap<>();
    String headerKey = "Content-Type";
    headers.put(headerKey, "text/plain");

    boolean result =
        api.openUrlInApp(
            url,
            true,
            new Messages.WebViewOptions.Builder()
                .setEnableJavaScript(false)
                .setEnableDomStorage(false)
                .setHeaders(headers)
                .build(),
            new Messages.BrowserOptions.Builder().setShowTitle(false).build());

    final ArgumentCaptor<Intent> intentCaptor = ArgumentCaptor.forClass(Intent.class);
    verify(activity).startActivity(intentCaptor.capture(), isNull());
    assertTrue(result);
    assertEquals(Intent.ACTION_VIEW, intentCaptor.getValue().getAction());
    assertNull(intentCaptor.getValue().getComponent());
    final Bundle passedHeaders =
        intentCaptor.getValue().getExtras().getBundle(Browser.EXTRA_HEADERS);
    assertEquals(headers.get(headerKey), passedHeaders.getString(headerKey));
  }

  @Test
  public void openWebView_opensUrlInCustomTabsWithShowTitle() {
    Activity activity = mock(Activity.class);
    UrlLauncher api = new UrlLauncher(ApplicationProvider.getApplicationContext());
    api.setActivity(activity);
    String url = "https://flutter.dev";
    HashMap<String, String> headers = new HashMap<>();

    boolean result =
        api.openUrlInApp(
            url,
            true,
            new Messages.WebViewOptions.Builder()
                .setEnableJavaScript(false)
                .setEnableDomStorage(false)
                .setHeaders(headers)
                .build(),
            new Messages.BrowserOptions.Builder().setShowTitle(true).build());

    final ArgumentCaptor<Intent> intentCaptor = ArgumentCaptor.forClass(Intent.class);
    verify(activity).startActivity(intentCaptor.capture(), isNull());
    assertTrue(result);
    assertEquals(Intent.ACTION_VIEW, intentCaptor.getValue().getAction());
    assertNull(intentCaptor.getValue().getComponent());
    assertEquals(
        intentCaptor.getValue().getExtras().getInt(CustomTabsIntent.EXTRA_TITLE_VISIBILITY_STATE),
        CustomTabsIntent.SHOW_PAGE_TITLE);
  }

  @Test
  public void openWebView_opensUrlInCustomTabsWithoutShowTitle() {
    Activity activity = mock(Activity.class);
    UrlLauncher api = new UrlLauncher(ApplicationProvider.getApplicationContext());
    api.setActivity(activity);
    String url = "https://flutter.dev";
    HashMap<String, String> headers = new HashMap<>();

    boolean result =
        api.openUrlInApp(
            url,
            true,
            new Messages.WebViewOptions.Builder()
                .setEnableJavaScript(false)
                .setEnableDomStorage(false)
                .setHeaders(headers)
                .build(),
            new Messages.BrowserOptions.Builder().setShowTitle(false).build());

    final ArgumentCaptor<Intent> intentCaptor = ArgumentCaptor.forClass(Intent.class);
    verify(activity).startActivity(intentCaptor.capture(), isNull());
    assertTrue(result);
    assertEquals(Intent.ACTION_VIEW, intentCaptor.getValue().getAction());
    assertNull(intentCaptor.getValue().getComponent());
    assertEquals(
        intentCaptor.getValue().getExtras().getInt(CustomTabsIntent.EXTRA_TITLE_VISIBILITY_STATE),
        CustomTabsIntent.NO_TITLE);
  }

  @Test
  public void openWebView_fallsBackToWebViewIfCustomTabFails() {
    Activity activity = mock(Activity.class);
    UrlLauncher api = new UrlLauncher(ApplicationProvider.getApplicationContext());
    api.setActivity(activity);
    String url = "https://flutter.dev";
    doThrow(new ActivityNotFoundException())
        .when(activity)
        .startActivity(any(), isNull()); // for custom tabs intent

    boolean result =
        api.openUrlInApp(
            url,
            true,
            new Messages.WebViewOptions.Builder()
                .setEnableJavaScript(false)
                .setEnableDomStorage(false)
                .setHeaders(new HashMap<>())
                .build(),
            new Messages.BrowserOptions.Builder().setShowTitle(false).build());

    final ArgumentCaptor<Intent> intentCaptor = ArgumentCaptor.forClass(Intent.class);
    verify(activity).startActivity(intentCaptor.capture());
    assertTrue(result);
    assertEquals(url, intentCaptor.getValue().getExtras().getString(WebViewActivity.URL_EXTRA));
    assertEquals(
        false, intentCaptor.getValue().getExtras().getBoolean(WebViewActivity.ENABLE_JS_EXTRA));
    assertEquals(
        false, intentCaptor.getValue().getExtras().getBoolean(WebViewActivity.ENABLE_DOM_EXTRA));
  }

  @Test
  public void openWebView_handlesEnableJavaScript() {
    Activity activity = mock(Activity.class);
    UrlLauncher api = new UrlLauncher(ApplicationProvider.getApplicationContext());
    api.setActivity(activity);
    boolean enableJavaScript = true;
    HashMap<String, String> headers = new HashMap<>();
    headers.put("key", "value");

    api.openUrlInApp(
        "https://flutter.dev",
        true,
        new Messages.WebViewOptions.Builder()
            .setEnableJavaScript(enableJavaScript)
            .setEnableDomStorage(false)
            .setHeaders(headers)
            .build(),
        new Messages.BrowserOptions.Builder().setShowTitle(false).build());

    final ArgumentCaptor<Intent> intentCaptor = ArgumentCaptor.forClass(Intent.class);
    verify(activity).startActivity(intentCaptor.capture());
    assertEquals(
        enableJavaScript,
        intentCaptor.getValue().getExtras().getBoolean(WebViewActivity.ENABLE_JS_EXTRA));
  }

  @Test
  public void openWebView_handlesHeaders() {
    Activity activity = mock(Activity.class);
    UrlLauncher api = new UrlLauncher(ApplicationProvider.getApplicationContext());
    api.setActivity(activity);
    HashMap<String, String> headers = new HashMap<>();
    final String key1 = "key";
    final String key2 = "key2";
    headers.put(key1, "value");
    headers.put(key2, "value2");

    api.openUrlInApp(
        "https://flutter.dev",
        true,
        new Messages.WebViewOptions.Builder()
            .setEnableJavaScript(false)
            .setEnableDomStorage(false)
            .setHeaders(headers)
            .build(),
        new Messages.BrowserOptions.Builder().setShowTitle(false).build());

    final ArgumentCaptor<Intent> intentCaptor = ArgumentCaptor.forClass(Intent.class);
    verify(activity).startActivity(intentCaptor.capture());
    final Bundle passedHeaders =
        intentCaptor.getValue().getExtras().getBundle(Browser.EXTRA_HEADERS);
    assertEquals(headers.size(), passedHeaders.size());
    assertEquals(headers.get(key1), passedHeaders.getString(key1));
    assertEquals(headers.get(key2), passedHeaders.getString(key2));
  }

  @Test
  public void openWebView_handlesEnableDomStorage() {
    Activity activity = mock(Activity.class);
    UrlLauncher api = new UrlLauncher(ApplicationProvider.getApplicationContext());
    api.setActivity(activity);
    boolean enableDomStorage = true;
    HashMap<String, String> headers = new HashMap<>();
    headers.put("key", "value");

    api.openUrlInApp(
        "https://flutter.dev",
        true,
        new Messages.WebViewOptions.Builder()
            .setEnableJavaScript(false)
            .setEnableDomStorage(enableDomStorage)
            .setHeaders(headers)
            .build(),
        new Messages.BrowserOptions.Builder().setShowTitle(false).build());

    final ArgumentCaptor<Intent> intentCaptor = ArgumentCaptor.forClass(Intent.class);
    verify(activity).startActivity(intentCaptor.capture());
    assertEquals(
        enableDomStorage,
        intentCaptor.getValue().getExtras().getBoolean(WebViewActivity.ENABLE_DOM_EXTRA));
  }

  @Test
  public void openWebView_handlesEnableShowTitle() {
    Activity activity = mock(Activity.class);
    UrlLauncher api = new UrlLauncher(ApplicationProvider.getApplicationContext());
    api.setActivity(activity);
    boolean enableDomStorage = true;
    HashMap<String, String> headers = new HashMap<>();
    boolean showTitle = true;

    api.openUrlInApp(
        "https://flutter.dev",
        true,
        new Messages.WebViewOptions.Builder()
            .setEnableJavaScript(false)
            .setEnableDomStorage(enableDomStorage)
            .setHeaders(headers)
            .build(),
        new Messages.BrowserOptions.Builder().setShowTitle(showTitle).build());

    final ArgumentCaptor<Intent> intentCaptor = ArgumentCaptor.forClass(Intent.class);
    verify(activity).startActivity(intentCaptor.capture(), isNull());

    assertEquals(
        intentCaptor.getValue().getExtras().getInt(CustomTabsIntent.EXTRA_TITLE_VISIBILITY_STATE),
        CustomTabsIntent.SHOW_PAGE_TITLE);
  }

  @Test
  public void openWebView_throwsForNoCurrentActivity() {
    UrlLauncher api = new UrlLauncher(ApplicationProvider.getApplicationContext());
    api.setActivity(null);

    Messages.FlutterError exception =
        assertThrows(
            Messages.FlutterError.class,
            () ->
                api.openUrlInApp(
                    "https://flutter.dev",
                    true,
                    new Messages.WebViewOptions.Builder()
                        .setEnableJavaScript(false)
                        .setEnableDomStorage(false)
                        .setHeaders(new HashMap<>())
                        .build(),
                    new Messages.BrowserOptions.Builder().setShowTitle(false).build()));
    assertEquals("NO_ACTIVITY", exception.code);
  }

  @Test
  public void openWebView_returnsFalse() {
    Activity activity = mock(Activity.class);
    UrlLauncher api = new UrlLauncher(ApplicationProvider.getApplicationContext());
    api.setActivity(activity);
    doThrow(new ActivityNotFoundException())
        .when(activity)
        .startActivity(any(), isNull()); // for custom tabs intent
    doThrow(new ActivityNotFoundException())
        .when(activity)
        .startActivity(any()); // for webview intent

    boolean result =
        api.openUrlInApp(
            "https://flutter.dev",
            true,
            new Messages.WebViewOptions.Builder()
                .setEnableJavaScript(false)
                .setEnableDomStorage(false)
                .setHeaders(new HashMap<>())
                .build(),
            new Messages.BrowserOptions.Builder().setShowTitle(false).build());

    assertFalse(result);
  }

  @Test
  public void closeWebView_closes() {
    Context context = mock(Context.class);
    UrlLauncher api = new UrlLauncher(context);

    api.closeWebView();

    final ArgumentCaptor<Intent> intentCaptor = ArgumentCaptor.forClass(Intent.class);
    verify(context).sendBroadcast(intentCaptor.capture());
    assertEquals(WebViewActivity.ACTION_CLOSE, intentCaptor.getValue().getAction());
  }
}

File: packages/url_launcher/url_launcher_android/android/src/test/kotlin/io/flutter/plugins/urllauncher/UrlLauncherTest.kt
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
package io.flutter.plugins.urllauncher

import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.provider.Browser
import androidx.browser.customtabs.CustomTabsIntent
import androidx.test.core.app.ApplicationProvider
import io.flutter.plugins.urllauncher.UrlLauncher.IntentResolver
import org.junit.Assert
import org.junit.Test
import org.junit.function.ThrowingRunnable
import org.junit.runner.RunWith
import org.mockito.ArgumentCaptor
import org.mockito.ArgumentMatchers
import org.mockito.Mockito
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

@RunWith(RobolectricTestRunner::class)
class UrlLauncherTest {
  @Test
  fun canLaunch_createsIntentWithPassedUrl() {
    val resolver = Mockito.mock<IntentResolver>(IntentResolver::class.java)
    val api = UrlLauncher(ApplicationProvider.getApplicationContext<Context?>(), resolver)
    val url = Uri.parse("https://flutter.dev")
    Mockito.`when`<String?>(resolver.getHandlerComponentName(ArgumentMatchers.any<Intent?>()))
        .thenReturn(null)

    api.canLaunchUrl(url.toString())

    val intentCaptor = ArgumentCaptor.forClass<Intent?, Intent?>(Intent::class.java)
    Mockito.verify<IntentResolver?>(resolver).getHandlerComponentName(intentCaptor.capture()!!)
    Assert.assertEquals(url, intentCaptor.getValue()!!.getData())
  }

  @Test
  fun canLaunch_returnsTrue() {
    val api =
        UrlLauncher(
            ApplicationProvider.getApplicationContext<Context?>(),
            IntentResolver { intent: Intent? -> "some.component" })

    val result = api.canLaunchUrl("https://flutter.dev")

    Assert.assertTrue(result)
  }

  @Test
  fun canLaunch_returnsFalse() {
    val api =
        UrlLauncher(
            ApplicationProvider.getApplicationContext<Context?>(),
            IntentResolver { intent: Intent? -> null })

    val result = api.canLaunchUrl("https://flutter.dev")

    Assert.assertFalse(result)
  }

  // Integration testing on emulators won't work as expected without the workaround this tests
  // for, since it will be returned even for intentionally bogus schemes.
  @Test
  fun canLaunch_returnsFalseForEmulatorFallbackComponent() {
    val api =
        UrlLauncher(
            ApplicationProvider.getApplicationContext<Context?>(),
            IntentResolver { intent: Intent? ->
              "{com.android.fallback/com.android.fallback.Fallback}"
            })

    val result = api.canLaunchUrl("https://flutter.dev")

    Assert.assertFalse(result)
  }

  @Test
  fun launch_throwsForNoCurrentActivity() {
    val api = UrlLauncher(ApplicationProvider.getApplicationContext<Context?>())
    api.setActivity(null)

    val exception =
        Assert.assertThrows<FlutterError>(
            FlutterError::class.java,
            ThrowingRunnable {
              api.launchUrl("https://flutter.dev", HashMap<String?, String?>(), false)
            })
    Assert.assertEquals("NO_ACTIVITY", exception.code)
  }

  @Test
  fun launch_createsIntentWithPassedUrl() {
    val activity = Mockito.mock<Activity?>(Activity::class.java)
    val url = "https://flutter.dev"
    val api = UrlLauncher(ApplicationProvider.getApplicationContext<Context?>())
    api.setActivity(activity)
    Mockito.doThrow(ActivityNotFoundException())
        .`when`<Activity?>(activity)
        .startActivity(ArgumentMatchers.any<Intent?>())

    api.launchUrl("https://flutter.dev", HashMap<String?, String?>(), false)

    val intentCaptor = ArgumentCaptor.forClass<Intent?, Intent?>(Intent::class.java)
    Mockito.verify<Activity?>(activity).startActivity(intentCaptor.capture())
    Assert.assertEquals(url, intentCaptor.getValue()!!.getData().toString())
    Assert.assertEquals(
        0,
        (intentCaptor.getValue()!!.getFlags() and Intent.FLAG_ACTIVITY_REQUIRE_NON_BROWSER)
            .toLong())
  }

  @Config(minSdk = 30)
  @Test
  fun launch_setsRequireNonBrowserWhenRequested() {
    val activity = Mockito.mock<Activity?>(Activity::class.java)
    val api = UrlLauncher(ApplicationProvider.getApplicationContext<Context?>())
    api.setActivity(activity)
    Mockito.doThrow(ActivityNotFoundException())
        .`when`<Activity?>(activity)
        .startActivity(ArgumentMatchers.any<Intent?>())

    api.launchUrl("https://flutter.dev", HashMap<String?, String?>(), true)

    val intentCaptor = ArgumentCaptor.forClass<Intent?, Intent?>(Intent::class.java)
    Mockito.verify<Activity?>(activity).startActivity(intentCaptor.capture())
    Assert.assertEquals(
        Intent.FLAG_ACTIVITY_REQUIRE_NON_BROWSER.toLong(),
        (intentCaptor.getValue()!!.getFlags() and Intent.FLAG_ACTIVITY_REQUIRE_NON_BROWSER)
            .toLong())
  }

  @Test
  fun launch_returnsFalse() {
    val activity = Mockito.mock<Activity?>(Activity::class.java)
    val api = UrlLauncher(ApplicationProvider.getApplicationContext<Context?>())
    api.setActivity(activity)
    Mockito.doThrow(ActivityNotFoundException())
        .`when`<Activity?>(activity)
        .startActivity(ArgumentMatchers.any<Intent?>())

    val result = api.launchUrl("https://flutter.dev", HashMap<String?, String?>(), false)

    Assert.assertFalse(result)
  }

  @Test
  fun launch_returnsTrue() {
    val activity = Mockito.mock<Activity?>(Activity::class.java)
    val api = UrlLauncher(ApplicationProvider.getApplicationContext<Context?>())
    api.setActivity(activity)

    val result = api.launchUrl("https://flutter.dev", HashMap<String?, String?>(), false)

    Assert.assertTrue(result)
  }

  @Test
  fun openUrlInApp_opensUrlInWebViewIfNecessary() {
    val activity = Mockito.mock<Activity?>(Activity::class.java)
    val api = UrlLauncher(ApplicationProvider.getApplicationContext<Context?>())
    api.setActivity(activity)
    val url = "https://flutter.dev"
    val enableJavaScript = false
    val enableDomStorage = false
    val headers = HashMap<String?, String?>()
    headers.put("key", "value")
    val showTitle = false

    val result =
        api.openUrlInApp(
            url,
            true,
            WebViewOptions(enableJavaScript, enableDomStorage, headers),
            BrowserOptions(showTitle))

    val intentCaptor = ArgumentCaptor.forClass<Intent?, Intent?>(Intent::class.java)
    Mockito.verify<Activity?>(activity).startActivity(intentCaptor.capture())
    Assert.assertTrue(result)
    Assert.assertEquals(
        url, intentCaptor.getValue()!!.getExtras()!!.getString(WebViewActivity.URL_EXTRA))
    Assert.assertEquals(
        enableJavaScript,
        intentCaptor.getValue()!!.getExtras()!!.getBoolean(WebViewActivity.ENABLE_JS_EXTRA))
    Assert.assertEquals(
        enableDomStorage,
        intentCaptor.getValue()!!.getExtras()!!.getBoolean(WebViewActivity.ENABLE_DOM_EXTRA))
  }

  @Test
  fun openWebView_opensUrlInWebViewIfRequested() {
    val activity = Mockito.mock<Activity?>(Activity::class.java)
    val api = UrlLauncher(ApplicationProvider.getApplicationContext<Context?>())
    api.setActivity(activity)
    val url = "https://flutter.dev"

    val result =
        api.openUrlInApp(
            url,
            false,
            WebViewOptions(false, false, HashMap<String?, String?>()),
            BrowserOptions(true))

    val intentCaptor = ArgumentCaptor.forClass<Intent?, Intent?>(Intent::class.java)
    Mockito.verify<Activity?>(activity).startActivity(intentCaptor.capture())
    Assert.assertTrue(result)
    Assert.assertEquals(
        url, intentCaptor.getValue()!!.getExtras()!!.getString(WebViewActivity.URL_EXTRA))
  }

  @Test
  fun openWebView_opensUrlInCustomTabs() {
    val activity = Mockito.mock<Activity?>(Activity::class.java)
    val api = UrlLauncher(ApplicationProvider.getApplicationContext<Context?>())
    api.setActivity(activity)
    val url = "https://flutter.dev"

    val result =
        api.openUrlInApp(
            url,
            true,
            WebViewOptions(false, false, HashMap<String?, String?>()),
            BrowserOptions(false))

    val intentCaptor = ArgumentCaptor.forClass<Intent?, Intent?>(Intent::class.java)
    Mockito.verify<Activity?>(activity)
        .startActivity(intentCaptor.capture(), ArgumentMatchers.any<Bundle?>())
    Assert.assertTrue(result)
    Assert.assertEquals(Intent.ACTION_VIEW, intentCaptor.getValue()!!.getAction())
    Assert.assertNull(intentCaptor.getValue()!!.getComponent())
  }

  @Test
  fun openWebView_opensUrlInCustomTabsWithCORSAllowedHeader() {
    val activity = Mockito.mock<Activity?>(Activity::class.java)
    val api = UrlLauncher(ApplicationProvider.getApplicationContext<Context?>())
    api.setActivity(activity)
    val url = "https://flutter.dev"
    val headers = HashMap<String?, String?>()
    val headerKey = "Content-Type"
    headers.put(headerKey, "text/plain")

    val result =
        api.openUrlInApp(url, true, WebViewOptions(false, false, headers), BrowserOptions(false))

    val intentCaptor = ArgumentCaptor.forClass<Intent?, Intent?>(Intent::class.java)
    Mockito.verify<Activity?>(activity)
        .startActivity(intentCaptor.capture(), ArgumentMatchers.any<Bundle?>())
    Assert.assertTrue(result)
    Assert.assertEquals(Intent.ACTION_VIEW, intentCaptor.getValue()!!.getAction())
    Assert.assertNull(intentCaptor.getValue()!!.getComponent())
    val passedHeaders = intentCaptor.getValue()!!.getExtras()!!.getBundle(Browser.EXTRA_HEADERS)
    Assert.assertEquals(headers.get(headerKey), passedHeaders!!.getString(headerKey))
  }

  @Test
  fun openWebView_opensUrlInCustomTabsWithShowTitle() {
    val activity = Mockito.mock<Activity?>(Activity::class.java)
    val api = UrlLauncher(ApplicationProvider.getApplicationContext<Context?>())
    api.setActivity(activity)
    val url = "https://flutter.dev"
    val headers = HashMap<String?, String?>()

    val result =
        api.openUrlInApp(url, true, WebViewOptions(false, false, headers), BrowserOptions(true))

    val intentCaptor = ArgumentCaptor.forClass<Intent?, Intent?>(Intent::class.java)
    Mockito.verify<Activity?>(activity)
        .startActivity(intentCaptor.capture(), ArgumentMatchers.any<Bundle?>())
    Assert.assertTrue(result)
    Assert.assertEquals(Intent.ACTION_VIEW, intentCaptor.getValue()!!.getAction())
    Assert.assertNull(intentCaptor.getValue()!!.getComponent())
    Assert.assertEquals(
        CustomTabsIntent.SHOW_PAGE_TITLE.toLong(),
        intentCaptor
            .getValue()!!
            .getExtras()!!
            .getInt(CustomTabsIntent.EXTRA_TITLE_VISIBILITY_STATE)
            .toLong())
  }

  @Test
  fun openWebView_opensUrlInCustomTabsWithoutShowTitle() {
    val activity = Mockito.mock<Activity?>(Activity::class.java)
    val api = UrlLauncher(ApplicationProvider.getApplicationContext<Context?>())
    api.setActivity(activity)
    val url = "https://flutter.dev"
    val headers = HashMap<String?, String?>()

    val result =
        api.openUrlInApp(url, true, WebViewOptions(false, false, headers), BrowserOptions(false))

    val intentCaptor = ArgumentCaptor.forClass<Intent?, Intent?>(Intent::class.java)
    Mockito.verify<Activity?>(activity)
        .startActivity(intentCaptor.capture(), ArgumentMatchers.any<Bundle?>())
    Assert.assertTrue(result)
    Assert.assertEquals(Intent.ACTION_VIEW, intentCaptor.getValue()!!.getAction())
    Assert.assertNull(intentCaptor.getValue()!!.getComponent())
    Assert.assertEquals(
        CustomTabsIntent.NO_TITLE.toLong(),
        intentCaptor
            .getValue()!!
            .getExtras()!!
            .getInt(CustomTabsIntent.EXTRA_TITLE_VISIBILITY_STATE)
            .toLong())
  }

  @Test
  fun openWebView_fallsBackToWebViewIfCustomTabFails() {
    val activity = Mockito.mock<Activity?>(Activity::class.java)
    val api = UrlLauncher(ApplicationProvider.getApplicationContext<Context?>())
    api.setActivity(activity)
    val url = "https://flutter.dev"
    Mockito.doThrow(ActivityNotFoundException())
        .`when`<Activity?>(activity)
        .startActivity(
            ArgumentMatchers.any<Intent?>(),
            ArgumentMatchers.any<Bundle?>()) // for custom tabs intent

    val result =
        api.openUrlInApp(
            url,
            true,
            WebViewOptions(false, false, HashMap<String?, String?>()),
            BrowserOptions(false))

    val intentCaptor = ArgumentCaptor.forClass<Intent?, Intent?>(Intent::class.java)
    Mockito.verify<Activity?>(activity).startActivity(intentCaptor.capture())
    Assert.assertTrue(result)
    Assert.assertEquals(
        url, intentCaptor.getValue()!!.getExtras()!!.getString(WebViewActivity.URL_EXTRA))
    Assert.assertFalse(
        intentCaptor.getValue()!!.getExtras()!!.getBoolean(WebViewActivity.ENABLE_JS_EXTRA))
    Assert.assertFalse(
        intentCaptor.getValue()!!.getExtras()!!.getBoolean(WebViewActivity.ENABLE_DOM_EXTRA))
  }

  @Test
  fun openWebView_handlesEnableJavaScript() {
    val activity = Mockito.mock<Activity?>(Activity::class.java)
    val api = UrlLauncher(ApplicationProvider.getApplicationContext<Context?>())
    api.setActivity(activity)
    val enableJavaScript = true
    val headers = HashMap<String?, String?>()
    headers.put("key", "value")

    api.openUrlInApp(
        "https://flutter.dev",
        true,
        WebViewOptions(enableJavaScript, false, headers),
        BrowserOptions(false))

    val intentCaptor = ArgumentCaptor.forClass<Intent?, Intent?>(Intent::class.java)
    Mockito.verify<Activity?>(activity).startActivity(intentCaptor.capture())
    Assert.assertEquals(
        enableJavaScript,
        intentCaptor.getValue()!!.getExtras()!!.getBoolean(WebViewActivity.ENABLE_JS_EXTRA))
  }

  @Test
  fun openWebView_handlesHeaders() {
    val activity = Mockito.mock<Activity?>(Activity::class.java)
    val api = UrlLauncher(ApplicationProvider.getApplicationContext<Context?>())
    api.setActivity(activity)
    val headers = HashMap<String?, String?>()
    val key1 = "key"
    val key2 = "key2"
    headers.put(key1, "value")
    headers.put(key2, "value2")

    api.openUrlInApp(
        "https://flutter.dev", true, WebViewOptions(false, false, headers), BrowserOptions(false))

    val intentCaptor = ArgumentCaptor.forClass<Intent?, Intent?>(Intent::class.java)
    Mockito.verify<Activity?>(activity).startActivity(intentCaptor.capture())
    val passedHeaders = intentCaptor.getValue()!!.getExtras()!!.getBundle(Browser.EXTRA_HEADERS)
    Assert.assertEquals(headers.size.toLong(), passedHeaders!!.size().toLong())
    Assert.assertEquals(headers.get(key1), passedHeaders.getString(key1))
    Assert.assertEquals(headers.get(key2), passedHeaders.getString(key2))
  }

  @Test
  fun openWebView_handlesEnableDomStorage() {
    val activity = Mockito.mock<Activity?>(Activity::class.java)
    val api = UrlLauncher(ApplicationProvider.getApplicationContext<Context?>())
    api.setActivity(activity)
    val enableDomStorage = true
    val headers = HashMap<String?, String?>()
    headers.put("key", "value")

    api.openUrlInApp(
        "https://flutter.dev",
        true,
        WebViewOptions(false, enableDomStorage, headers),
        BrowserOptions(false))

    val intentCaptor = ArgumentCaptor.forClass<Intent?, Intent?>(Intent::class.java)
    Mockito.verify<Activity?>(activity).startActivity(intentCaptor.capture())
    Assert.assertEquals(
        enableDomStorage,
        intentCaptor.getValue()!!.getExtras()!!.getBoolean(WebViewActivity.ENABLE_DOM_EXTRA))
  }

  @Test
  fun openWebView_handlesEnableShowTitle() {
    val activity = Mockito.mock<Activity?>(Activity::class.java)
    val api = UrlLauncher(ApplicationProvider.getApplicationContext<Context?>())
    api.setActivity(activity)
    val enableDomStorage = true
    val headers = HashMap<String?, String?>()
    val showTitle = true

    api.openUrlInApp(
        "https://flutter.dev",
        true,
        WebViewOptions(false, enableDomStorage, headers),
        BrowserOptions(showTitle))

    val intentCaptor = ArgumentCaptor.forClass<Intent?, Intent?>(Intent::class.java)
    Mockito.verify<Activity?>(activity)
        .startActivity(intentCaptor.capture(), ArgumentMatchers.any<Bundle?>())

    Assert.assertEquals(
        CustomTabsIntent.SHOW_PAGE_TITLE.toLong(),
        intentCaptor
            .getValue()!!
            .getExtras()!!
            .getInt(CustomTabsIntent.EXTRA_TITLE_VISIBILITY_STATE)
            .toLong())
  }

  @Test
  fun openWebView_throwsForNoCurrentActivity() {
    val api = UrlLauncher(ApplicationProvider.getApplicationContext<Context?>())
    api.setActivity(null)

    val exception =
        Assert.assertThrows<FlutterError>(
            FlutterError::class.java,
            ThrowingRunnable {
              api.openUrlInApp(
                  "https://flutter.dev",
                  true,
                  WebViewOptions(false, false, HashMap<String?, String?>()),
                  BrowserOptions(false))
            })
    Assert.assertEquals("NO_ACTIVITY", exception.code)
  }

  @Test
  fun openWebView_returnsFalse() {
    val activity = Mockito.mock<Activity?>(Activity::class.java)
    val api = UrlLauncher(ApplicationProvider.getApplicationContext<Context?>())
    api.setActivity(activity)
    Mockito.doThrow(ActivityNotFoundException())
        .`when`<Activity?>(activity)
        .startActivity(
            ArgumentMatchers.any<Intent?>(),
            ArgumentMatchers.any<Bundle?>()) // for custom tabs intent
    Mockito.doThrow(ActivityNotFoundException())
        .`when`<Activity?>(activity)
        .startActivity(ArgumentMatchers.any<Intent?>()) // for webview intent

    val result =
        api.openUrlInApp(
            "https://flutter.dev",
            true,
            WebViewOptions(false, false, HashMap<String?, String?>()),
            BrowserOptions(false))

    Assert.assertFalse(result)
  }

  @Test
  fun closeWebView_closes() {
    val context = Mockito.mock<Context>(Context::class.java)
    val api = UrlLauncher(context)

    api.closeWebView()

    val intentCaptor = ArgumentCaptor.forClass<Intent?, Intent?>(Intent::class.java)
    Mockito.verify<Context?>(context).sendBroadcast(intentCaptor.capture())
    Assert.assertEquals(WebViewActivity.ACTION_CLOSE, intentCaptor.getValue()!!.getAction())
  }
}
